### PR TITLE
Aurora for postgres RDS engine support

### PR DESCRIPTION
**Pull request by miromad**
Making sure terraform supports Aurora PostgreSQL now that we're getting close to its GA.
- allows explicitly setting one of two supported engine types
  aurora (default) or aurora-postgresql
- adds optional engine_version which may be useful for pinning specific
  postgres engine version
- maintains backward compatibility with existing aurora rds clusters
  with combinaton of defaults and optional parameters
- used and tested under "Aurora PostgreSQL Preview" authorized aws account in us-east-1 region

```
github.com/terraform-providers/terraform-provider-aws$ TF_ACC=1 go test -v github.com/terraform-providers/terraform-provider-aws/aws -timeout=7200s -run "TestAccAWSRDSClusterInstance_.*"
=== RUN   TestAccAWSRDSClusterInstance_importBasic
--- PASS: TestAccAWSRDSClusterInstance_importBasic (770.69s)
=== RUN   TestAccAWSRDSClusterInstance_basic
--- PASS: TestAccAWSRDSClusterInstance_basic (1440.30s)
=== RUN   TestAccAWSRDSClusterInstance_namePrefix
--- PASS: TestAccAWSRDSClusterInstance_namePrefix (764.54s)
=== RUN   TestAccAWSRDSClusterInstance_generatedName
--- PASS: TestAccAWSRDSClusterInstance_generatedName (720.21s)
=== RUN   TestAccAWSRDSClusterInstance_kmsKey
--- PASS: TestAccAWSRDSClusterInstance_kmsKey (765.96s)
=== RUN   TestAccAWSRDSClusterInstance_disappears
--- PASS: TestAccAWSRDSClusterInstance_disappears (722.53s)
=== RUN   TestAccAWSRDSClusterInstance_withInstanceEnhancedMonitor
--- PASS: TestAccAWSRDSClusterInstance_withInstanceEnhancedMonitor (735.03s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	5919.258s

github.com/terraform-providers/terraform-provider-aws$ TF_ACC=1 go test -v github.com/terraform-providers/terraform-provider-aws/aws -timeout=7200s -run "TestAccAWSRDSCluster_.*"
=== RUN   TestAccAWSRDSCluster_importBasic
--- PASS: TestAccAWSRDSCluster_importBasic (155.95s)
=== RUN   TestAccAWSRDSCluster_basic
--- PASS: TestAccAWSRDSCluster_basic (119.60s)
=== RUN   TestAccAWSRDSCluster_namePrefix
--- PASS: TestAccAWSRDSCluster_namePrefix (185.24s)
=== RUN   TestAccAWSRDSCluster_generatedName
--- PASS: TestAccAWSRDSCluster_generatedName (196.65s)
=== RUN   TestAccAWSRDSCluster_takeFinalSnapshot
--- PASS: TestAccAWSRDSCluster_takeFinalSnapshot (197.50s)
=== RUN   TestAccAWSRDSCluster_missingUserNameCausesError
--- PASS: TestAccAWSRDSCluster_missingUserNameCausesError (9.30s)
=== RUN   TestAccAWSRDSCluster_updateTags
--- PASS: TestAccAWSRDSCluster_updateTags (175.84s)
=== RUN   TestAccAWSRDSCluster_updateIamRoles
--- PASS: TestAccAWSRDSCluster_updateIamRoles (189.80s)
=== RUN   TestAccAWSRDSCluster_kmsKey
--- PASS: TestAccAWSRDSCluster_kmsKey (195.33s)
=== RUN   TestAccAWSRDSCluster_encrypted
--- PASS: TestAccAWSRDSCluster_encrypted (152.58s)
=== RUN   TestAccAWSRDSCluster_backupsUpdate
--- PASS: TestAccAWSRDSCluster_backupsUpdate (165.99s)
=== RUN   TestAccAWSRDSCluster_iamAuth
--- PASS: TestAccAWSRDSCluster_iamAuth (141.66s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1885.459s
```

### CODE REVIEW

**Comment by jcomeaux** [comment]
Hi @miromad ....  I believe a very similar effort has already been merged? https://github.com/terraform-providers/terraform-provider-aws/pull/1415

Is the engine selection not working (e.g. setting `engine = aurora-postgresql`)

***also probably relevant: https://github.com/aws/aws-sdk-go/issues/1470

**Comment by miromad** [comment]
Hi @jcomeaux.
My changes are merged with #1415. It added engine as an option to instance but not to cluster where it was left hardcoded. I also added type validation to make it more user friendly. Regarding #1470, since I tested this under preview account and everything worked no changes are needed in SDK to support it.

File: aws/resource_aws_rds_cluster.go [hunk 1]
```diff
@@ -96,9 +96,18 @@ func resourceAwsRDSCluster() *schema.Resource {
 			},
 
 			"engine": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				Default:      "aurora",
+				ForceNew:     true,
+				ValidateFunc: validateRdsEngine,
```
Comment (Ninir) [inline aws/resource_aws_rds_cluster.go]: I don't think we should validate the engine value, as we would need to update the code each time a new engine is available. Thoughts?

File: aws/resource_aws_rds_cluster.go [hunk 2]
```diff
@@ -96,9 +96,18 @@ func resourceAwsRDSCluster() *schema.Resource {
...
+			},
+
+			"engine_version": {
```
Comment (Ninir) [inline aws/resource_aws_rds_cluster.go]: We should also read the engine version around line 575

File: aws/resource_aws_rds_cluster_instance.go [hunk 3]
```diff
@@ -292,6 +311,7 @@ func resourceAwsRDSClusterInstanceRead(d *schema.ResourceData, meta interface{})
 
 	d.Set("publicly_accessible", db.PubliclyAccessible)
 	d.Set("cluster_identifier", db.DBClusterIdentifier)
+	d.Set("engine", db.Engine)
```
Comment (Ninir) [inline aws/resource_aws_rds_cluster_instance.go]: We should also read the engine_version here

File: aws/resource_aws_rds_cluster.go (hunk 1 above)
Comment (miromad) [reply aws/resource_aws_rds_cluster.go]: Valid point however, amazon introduced aurora for MySQL in 2014 and added PostgreSQL in 2017. Don't have insights into their product plans but so far if they keep adding one every 3 years updating code won't be an issue. I think having validation is more useful/user friendly than flexibility.

File: aws/resource_aws_rds_cluster.go (hunk 2 above)
Comment (miromad) [reply aws/resource_aws_rds_cluster.go]: right, missed that one.

File: aws/resource_aws_rds_cluster_instance.go (hunk 3 above)
Comment (miromad) [reply aws/resource_aws_rds_cluster_instance.go]: will fix it too

**Comment by miromad** [comment]
Hi @Ninir,
Thanks for feedback. Will make changes (updating doc and engine_version sets in 2 places you identified) over the weekend. Let me know if anything else is needed.

File: aws/resource_aws_rds_cluster.go (hunk 1 above)
Comment (Ninir) [reply aws/resource_aws_rds_cluster.go]: Looks good to me :)

**Comment by phanindra2407** [comment]
Hi @Ninir  ,so does terraform supports the cluster creation in aurora postgresql

**Comment by Ninir** [comment]
Hi @phanindra2407 

It is still not globally available, so we all need to be more patient here! 😄

**Comment by phanindra2407** [comment]
Hi @Ninir , Just want to know if I can be able to create RDS PostgreSQL cluster with instances with terraform provider 1.0.0

**Comment by jude-pieries** [comment]
Aurora-postgres is GA starting 24/10/2017 and is available in the following regions us-east-1,2 , us-west-2 and  EU-west-1

**Comment by Ninir** [comment]
Thanks for letting us know @jude-pieries :) 

Here's the link: https://aws.amazon.com/fr/blogs/aws/now-available-amazon-aurora-with-postgresql-compatibility/

**Comment by atrepca** [comment]
Confirming this works now, adding `engine = "aurora-postgresql"` does the trick.

**Pull request merged**